core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java

```diff
@@ -2688,29 +3025,79 @@ private VirtualRecordCursorFactory generateSelectVirtualWithSubquery(QueryModel
-        switch (model.getSetOperationType()) {
-            case QueryModel.SET_OPERATION_UNION:
-                return generateUnionFactory(model, masterFactory, executionContext, slaveFactory, SET_UNION_CONSTRUCTOR);
-            case QueryModel.SET_OPERATION_UNION_ALL:
-                return generateUnionAllFactory(model, masterFactory, executionContext, slaveFactory);
-            case QueryModel.SET_OPERATION_EXCEPT:
-                return generateUnionFactory(model, masterFactory, executionContext, slaveFactory, SET_EXCEPT_CONSTRUCTOR);
-            case QueryModel.SET_OPERATION_INTERSECT:
-                return generateUnionFactory(model, masterFactory, executionContext, slaveFactory, SET_INTERSECT_CONSTRUCTOR);
-            default:
-                assert false;
-                return null;
+        final RecordCursorFactory factoryB = generateQuery0(model.getUnionModel(), executionContext, true);
+        try {
+            final RecordMetadata metadataA = factoryA.getMetadata();
+            final RecordMetadata metadataB = factoryB.getMetadata();
+            RecordMetadata setMetadata;
+            final boolean castIsRequired = checkIfSetCastIsRequired(metadataA, metadataB);
+
+            switch (model.getSetOperationType()) {
+                case SET_OPERATION_UNION:
+                    setMetadata = castIsRequired ? widenSetMetadata(metadataA, metadataB) : GenericRecordMetadata.removeTimestamp(metadataA);
+
+                    return generateUnionFactory(
+                            model,
+                            executionContext,
+                            factoryA,
+                            factoryB,
+                            castIsRequired ? generateCastFunctions(setMetadata, metadataA, model.getModelPosition()) : null,
+                            castIsRequired ? generateCastFunctions(setMetadata, metadataB, model.getUnionModel().getModelPosition()) : null,
+                            setMetadata,
+                            SET_UNION_CONSTRUCTOR
+                    );
+                case SET_OPERATION_UNION_ALL:
+                    setMetadata = castIsRequired ? widenSetMetadata(metadataA, metadataB) : GenericRecordMetadata.removeTimestamp(metadataA);
+                    return generateUnionAllFactory(
+                            model,
+                            executionContext,
+                            factoryA,
+                            factoryB,
+                            castIsRequired ? generateCastFunctions(setMetadata, metadataA, model.getModelPosition()) : null,
+                            castIsRequired ? generateCastFunctions(setMetadata, metadataB, model.getUnionModel().getModelPosition()) : null,
+                            setMetadata
+                    );
+                case SET_OPERATION_EXCEPT:
+                    setMetadata = GenericRecordMetadata.removeTimestamp(metadataA);
```

Comment:
Why timestamp is removed? I think in except main table will still have the same timestamp as before except

Reply:
done in https://github.com/questdb/questdb/pull/2123/commits/0106b8ae0880e31459908933fef1ecbdf2ed2bfb